### DISCUSSION TITLE
How to understand the layout mapping in MMA Traits？

### DISCUSSION THREAD

**Discussion by MARD1NO** [NVIDIA/cutlass#1381] [Q&A] [answered]
Recently I'm learning Cute PPT, and I was confused about the layout mapping in MMA Traits:
<img width="1169" alt="image" src="https://github.com/NVIDIA/cutlass/assets/42901638/3d4a2851-d764-43b3-8ae2-caa58305f7d5">

How does it mapping logical value id to corrdinate? 

For example, the LayoutA in SM80_8x8x4_F64F64F64F64_TN and SM80_16x8x8_F32F16F16F32_TN is same. 
<img width="1213" alt="image" src="https://github.com/NVIDIA/cutlass/assets/42901638/4f75057a-a78b-468e-89df-1cd8d36745a8">

In SM80_8x8x4_F64F64F64F64_TN, the map we flatten is like: 
```
tid:            0 1 2 3 ....
mapping: 0 8 16 24 1 9 17 25
```
thread1 will map to the position 8, and in 8x4 Colmajor layout, it will map to (0, 1), am I right?

<img width="1212" alt="image" src="https://github.com/NVIDIA/cutlass/assets/42901638/bd9718d9-ca1a-4ecc-92d1-c5348a035dd1">

but in second case SM80_16x8x8_F32F16F16F32_TN, I was confused about each thread need to take 4 fp16 value, how does it map?

**Comment by ccecka** [comment, marked as answer]
Looks like a slideware typo to me.

The following program
```cpp
#include <cute/tensor.hpp>
using namespace cute;

int main()
{
  using MMA = MMA_Traits<SM80_16x8x8_F32F16F16F32_TN>;
  print("ALayout: "); print(typename MMA::ALayout{}); print("\n");
  print("BLayout: "); print(typename MMA::BLayout{}); print("\n");
  print("CLayout: "); print(typename MMA::CLayout{}); print("\n");
}
```
prints
```
ALayout: ((_4,_8),(_2,_2)):((_32,_1),(_16,_8))
BLayout: ((_4,_8),_2):((_16,_1),_8)
CLayout: ((_4,_8),(_2,_2)):((_32,_1),(_16,_8))
```
which makes a lot more sense: `ALayout` has size-4 in the V-mode and `BLayout` has size-2 in the V-mode.

Looks like you've got the right idea, but check out the `MMA_Atom` documentation as well: 
https://github.com/NVIDIA/cutlass/blob/main/media/docs/cute/0t_mma_atom.md

**Reply by MARD1NO** [reply]
Thanks Cecka! I will check out the documentation :D

**Reply by thakkarV** [reply]
can you please tell me which slide deck this was so I can update it? :)

**Reply by MARD1NO** [reply]
I just found the slides in other's cute Blog, the link is: https://www.cs.utexas.edu/users/flame/BLISRetreat2023/slides/Thakkar_BLISRetreat2023.pdf

**Reply by wanghongyu2001** [reply]
Hello everyone, in the diagram, "V0" and "V1" represent what.

**Reply by ccecka** [reply]
The 0th value and the 1st value that each thread owns.

**Reply by wanghongyu2001** [reply]
ALayout: ((_4,_8),(_2,_2)):((_32,_1),(_16,_8))
BLayout: ((_4,_8),_2):((_16,_1),_8)
CLayout: ((_4,_8),(_2,_2)):((_32,_1),(_16,_8))
Hello, how do we correspond this output with the image above?

**Reply by ccecka** [reply]
That's in the documentation:
https://github.com/NVIDIA/cutlass/blob/main/media/docs/cute/0t_mma_atom.md

**Reply by thakkarV** [reply]
I fixed this in my slides for SIAM PP24 :)

**Reply by wanghongyu2001** [reply]
Hi, can you send me the correct link to the slides?

**Reply by thakkarV** [reply]
Not yet. They will be available after the talk

**Reply by lee-bin** [reply]
@thakkarV Are the fixed slides available now?
